File: NX1980_NX1988_NX1992_NX1996_NX2000/UGOPEN/SampleNXOpenApplications/.NET/NXOpenExamples/EX_Modl_CreateExtruded.cs
/*=============================================================================

                    Copyright (c) 2004 UGS PLM Solutions
                    Unpublished - All rights reserved

===============================================================================
File description: Example for Wrapped UFUNC (NXOpen.NET)
/*  */
using System;
using System.IO;
using System.Collections;
using NXOpen;
using NXOpen.UF;

namespace NetExample
{
    /// This example will create EX_Modl_CreateExtruded.prt, EX_Modl_CreateExtruded.log files. 
    /// This example extrudes lines and arcs along the z-axis. Three lines and two arcs are drawn. 
    /// The center line is a reference line only, and is not included in the objects to be extruded.
    /// The program prints the report to the log file called "EX_Modl_CreateExtruded.log".
    public class EX_Modl_CreateExtruded
    {
        private static FileStream fs; 
        private static StreamWriter w;
        public static UFSession theUfSession;
        private static Session theSession;


        //Constants
        public const double PI = 3.14159265358979324;

         public int Execute()
         {
             Tag UFPart;
             string part_name = "EX_Modl_CreateExtruded";
             int units =2; 
             string name;


             theUfSession.Part.New(part_name, units, out UFPart);
             theUfSession.Part.AskPartName(UFPart, out name);
             w.WriteLine("Loaded: " + name);

             double[] l1_endpt1 = {-1.4476606170268,2.3788333712953,0.0};
             double[] l1_endpt2 = { 1.5471310496399,2.2539419253757,0.0};
             double[] l2_endpt1 = {-1.4476606170268,1.2548103580189,0.0};
             double[] l2_endpt2 = { 1.5471310496399,1.3797018039385,0.0};
             double[] l3_endpt1 = {-1.4710981170268,1.8168218646571,0.0};
             double[] l3_endpt2 = { 1.5289018829732,1.8168218646571,0.0};
             double[] direction = {0.0,0.0,1.0};
             double[] ref_pt= new double[3];
             double arc1_start_ang = (PI/180)*87.611984536731;
             double arc1_end_ang   = (PI/180)*272.38801546327;
             double arc1_rad       = 0.5625;
             double arc2_start_ang = (PI/180)*272.38801546327;
             double arc2_end_ang   = (PI/180)*447.61198453673;
             double arc2_rad       = 0.4375;
             string taper_angle = "0.0";
             string[] limit1 = {"1.0", "1.25"};
             int i, count = 4;
             Tag[] objarray= new Tag[5];
             Tag wcs_tag, matrix_tag;
             Tag[] loop_list, features;
             UFCurve.Line line1 = new UFCurve.Line();
             UFCurve.Line line2 = new UFCurve.Line();
             UFCurve.Line line3 = new UFCurve.Line();
             UFCurve.Arc arc1 = new UFCurve.Arc();
             UFCurve.Arc arc2 = new UFCurve.Arc();

             line1.start_point = new double[3];
             line1.start_point[0] = l1_endpt1[0];
             line1.start_point[1] = l1_endpt1[1];
             line1.start_point[2] = l1_endpt1[2];
             line1.end_point = new double[3];
             line1.end_point[0]   = l1_endpt2[0];
             line1.end_point[1]   = l1_endpt2[1];
             line1.end_point[2]   = l1_endpt2[2];

             line2.start_point = new double[3];
             line2.start_point[0] = l2_endpt1[0];
             line2.start_point[1] = l2_endpt1[1];
             line2.start_point[2] = l2_endpt1[2];
             line2.end_point = new double[3];
             line2.end_point[0]   = l2_endpt2[0];
             line2.end_point[1]   = l2_endpt2[1];
             line2.end_point[2]   = l2_endpt2[2];
            
             line3.start_point = new double[3];
             line3.start_point[0] = l3_endpt1[0];
             line3.start_point[1] = l3_endpt1[1];
             line3.start_point[2] = l3_endpt1[2];
             line3.end_point = new double[3];
             line3.end_point[0]   = l3_endpt2[0];
             line3.end_point[1]   = l3_endpt2[1];
             line3.end_point[2]   = l3_endpt2[2];
            
             arc1.start_angle   = arc1_start_ang;
             arc1.end_angle     = arc1_end_ang;
             arc1.arc_center = new double[3];
             arc1.arc_center[0] = l3_endpt1[0];
             arc1.arc_center[1] = l3_endpt1[1];
             arc1.arc_center[2] = l3_endpt1[2];
             arc1.radius        = arc1_rad;
             arc2.start_angle   = arc2_start_ang;
             arc2.end_angle     = arc2_end_ang;
             arc2.arc_center = new double[3];
             arc2.arc_center[0] = l3_endpt2[0];
             arc2.arc_center[1] = l3_endpt2[1];
             arc2.arc_center[2] = l3_endpt2[2];
             arc2.radius        = arc2_rad;

             theUfSession.Curve.CreateLine(ref line1,out objarray[0]);
             theUfSession.Curve.CreateLine(ref line2,out objarray[1]);
             theUfSession.Curve.CreateLine(ref line3,out objarray[4]);
             theUfSession.Csys.AskWcs(out wcs_tag);
             theUfSession.Csys.AskMatrixOfObject(wcs_tag, out matrix_tag);
             arc1.matrix_tag = matrix_tag;
             arc2.matrix_tag = matrix_tag;
             theUfSession.Curve.CreateArc(ref arc1,out objarray[2]);
             theUfSession.Curve.CreateArc(ref arc2, out objarray[3]);
             //Create loop list
             ArrayList arr_loop_list = new ArrayList();
             for(i = 0; i < count; i++) 
             {
                 arr_loop_list.Add(objarray[i]);
             }
             loop_list = (Tag [])arr_loop_list.ToArray(typeof(Tag));

             // while sending parameters to this function the order was 
             // loop_list, taper_angle, limit1, ref_pt, FeatureSigns.Nullsign, direction, out features
             // as above .. there were some compilation problem due to function prototype match.
             // Atul changed the order as follows in order to compile it successfully.
             // loop_list, taper_angle, limit1, ref_pt, direction, FeatureSigns.Nullsign, out features
             theUfSession.Modl.CreateExtruded(loop_list, taper_angle, limit1,
                 ref_pt, direction, FeatureSigns.Nullsign,  out features);

             theUfSession.Part.Save();
             return 0;
        }

        public static void Main(string[] args)
        {
            theSession=Session.GetSession();
            theUfSession= UFSession.GetUFSession();

            fs = new FileStream("EX_Modl_CreateExtruded.log", FileMode.Create, FileAccess.Write);
            w = new StreamWriter(fs); // create a stream writer 
            w.Write("Log Entry : \r\n");
            w.WriteLine("--Log entry goes here--"); 
            w.Flush(); // update underlying file

            if ( File.Exists("EX_Modl_CreateExtruded.prt") ) 
            {
                w.WriteLine("Remove EX_Modl_CreateExtruded.prt file from <Project Folder>\\bin\\Debug !!");
                w.WriteLine("EX_Modl_CreateExtruded.prt already exists. !!");
                w.Close();
                return;
            }

            try
            {
                EX_Modl_CreateExtruded curveTest1 = new EX_Modl_CreateExtruded();
                if (curveTest1.Execute()==0)
                {
                    w.WriteLine("Successful");
                }
                else
                {
                    w.WriteLine("Failed");
                }
            }
            catch(NXException e)
            {   
                w.WriteLine("Exception is: {0}", e.Message);
            }
            w.WriteLine("End of Log File");
            w.Close();
        }
        public static int GetUnloadOption(string dummy) 
        {
            return UFConstants.UF_UNLOAD_IMMEDIATELY;
        }
    }
}

File: NX8.5/NX8.5_32_8.5.3.3/UGOPEN/SampleNXOpenApplications/.NET/NXOpenExamples/EX_Modl_CreateExtruded.vb
' ==================================================================================================
'
'            Copyright (c) 2004 UGS PLM Solutions.
'                   Unpublished - All rights reserved
'
'====================================================================================================
'File description: Example for Wrapped UFUNC (NXOpen.NET)
' 
'====================================================================================================
Option Strict Off
Imports System
Imports System.IO
Imports System.Runtime.InteropServices
Imports NXOpen
Imports NXOpen.Features
Imports NXOpen.Preferences
Imports NXOpen.UF

Module EX_Modl_CreateExtruded

    Dim theSession As Session
    Dim theUfSession As UFSession

    Public Const PI As Double = 3.1415926535897931

    Public Sub Main()
        theSession = Session.GetSession()
        theUfSession = UFSession.GetUFSession()

        Dim fs As New FileStream("EX_Modl_CreateExtruded.log", FileMode.Create, FileAccess.Write)
        Dim stream As New StreamWriter(fs) ' create a stream writer 
        stream.WriteLine("--Log entry goes here--")
        stream.Flush() ' update underlying file

        If File.Exists("EX_Modl_CreateExtruded.prt") Then
            stream.WriteLine("Remove EX_Modl_CreateExtruded.prt file from <Project Folder>\\bin\\Debug !!")
            stream.WriteLine("EX_Modl_CreateExtruded.prt already exists. !!")
            stream.Close()
            Return
        End If

        Dim nError As Integer = Execute(stream, theUfSession)
        stream.WriteLine("End of Log File")
        stream.Close()

    End Sub

    Function Execute(ByVal stream As StreamWriter, ByVal theUfSession As NXOpen.UF.UFSession) As Integer
        Try
            Dim UFPart As NXOpen.Tag
            Dim part_name As String = "EX_Modl_CreateExtruded"
            Dim units As Integer = 2
            Dim name As String

            theUfSession.Part.[New](part_name, units, UFPart)
            stream.WriteLine("New part : ", part_name, " Created")

            Dim l1_endpt1() As Double = {-1.4476606170268, 2.3788333712953, 0.0}
            Dim l1_endpt2() As Double = {1.5471310496399, 2.2539419253757, 0.0}
            Dim l2_endpt1() As Double = {-1.4476606170268, 1.2548103580189, 0.0}
            Dim l2_endpt2() As Double = {1.5471310496399, 1.3797018039385, 0.0}
            Dim l3_endpt1() As Double = {-1.4710981170268, 1.8168218646571, 0.0}
            Dim l3_endpt2() As Double = {1.5289018829732, 1.8168218646571, 0.0}
            Dim direction() As Double = {0.0, 0.0, 1.0}
            Dim ref_pt(3) As Double
            Dim arc1_start_ang As Double = (PI / 180) * 87.611984536731
            Dim arc1_end_ang As Double = (PI / 180) * 272.38801546327
            Dim arc1_rad As Double = 0.5625
            Dim arc2_start_ang As Double = (PI / 180) * 272.38801546327
            Dim arc2_end_ang As Double = (PI / 180) * 447.61198453673
            Dim arc2_rad As Double = 0.4375
            Dim taper_angle As String = "0.0"
            Dim limit1() As String = {"1.0", "1.25"}
            Dim i As Integer
            Dim count As Integer = 4
            Dim objarray(5) As NXOpen.Tag
            Dim wcs_tag, matrix_tag As NXOpen.Tag
            Dim loop_list(), features() As NXOpen.Tag
            Dim line1 As NXOpen.UF.UFCurve.Line = New NXOpen.UF.UFCurve.Line
            Dim line2 As NXOpen.UF.UFCurve.Line = New NXOpen.UF.UFCurve.Line
            Dim line3 As NXOpen.UF.UFCurve.Line = New NXOpen.UF.UFCurve.Line
            Dim arc1 As NXOpen.UF.UFCurve.Arc = New NXOpen.UF.UFCurve.Arc
            Dim arc2 As NXOpen.UF.UFCurve.Arc = New NXOpen.UF.UFCurve.Arc

            line1.start_point = l1_endpt1
            line1.end_point = l1_endpt2
            line2.start_point = l2_endpt1
            line2.end_point = l2_endpt2
            line3.start_point = l3_endpt1
            line3.end_point = l3_endpt2

            arc1.start_angle = arc1_start_ang
            arc1.end_angle = arc1_end_ang
            arc1.arc_center = l3_endpt1
            arc1.radius = arc1_rad

            arc2.start_angle = arc2_start_ang
            arc2.end_angle = arc2_end_ang
            arc2.arc_center = l3_endpt2
            arc2.radius = arc2_rad

            theUfSession.Curve.CreateLine(line1, objarray(0))
            stream.WriteLine("Successful:Curve.CreateLine")

            theUfSession.Curve.CreateLine(line2, objarray(1))
            stream.WriteLine("Successful:Curve.CreateLine")

            theUfSession.Curve.CreateLine(line3, objarray(4))
            stream.WriteLine("Successful:Curve.CreateLine")

            theUfSession.Csys.AskWcs(wcs_tag)
            stream.WriteLine("Successful:Csys.AskWcs")

            theUfSession.Csys.AskMatrixOfObject(wcs_tag, matrix_tag)
            stream.WriteLine("Successful:Csys.AskMatrixOfObject")

            arc1.matrix_tag = matrix_tag
            arc2.matrix_tag = matrix_tag
            theUfSession.Curve.CreateArc(arc1, objarray(2))
            stream.WriteLine("Successful:Curve.CreateArc")

            theUfSession.Curve.CreateArc(arc2, objarray(3))
            stream.WriteLine("Successful:Curve.CreateArc")

            'Create loop list
            Dim arr_loop_list As System.Collections.ArrayList = New System.Collections.ArrayList
            For i = 0 To count - 1
                arr_loop_list.Add(objarray(i))
            Next
            Dim t As System.Type
            t = GetType(NXOpen.Tag)
            loop_list = arr_loop_list.ToArray(t)

            'while sending parameters to this function the order was 
            'loop_list, taper_angle, limit1, ref_pt, FeatureSigns.Nullsign, direction, out features
            'as above .. there were some compilation problem due to function prototype match.
            'Atul changed the order as follows in order to compile it successfully.
            'loop_list, taper_angle, limit1, ref_pt, direction, FeatureSigns.Nullsign, out features
            theUfSession.Modl.CreateExtruded(loop_list, taper_angle, limit1, ref_pt, direction, FeatureSigns.Nullsign, features)
            stream.WriteLine("Successful:Modl.CreateExtruded")

            'Save the Part
            theUfSession.Part.Save()
        Catch ex As Exception
            stream.WriteLine("EXCEPTION: ", ex.Message)
            Return 1
        End Try

        Return 0
    End Function

End Module


